### PR TITLE
Supporting for windows

### PR DESCRIPTION
**Pull request by sanzenwin**

### CODE REVIEW

File: scrapy_playwright/handler.py [hunk 1]
```diff
@@ -39,15 +40,50 @@
...
 
+# Supporting for Windows
+if sys.platform == "win32" and sys.version_info >= (3, 8):
+    import threading
+
+    class Var:
+        windows_loop = None
+        windows_thread = None
+
+    def windows_get_asyncio_event_loop():
```
Comment (elacuesta) [inline scrapy_playwright/handler.py:51]: Please make these functions underscored, ideally only the download handler should be public.

File: tests/__init__.py [hunk 2]
```diff
@@ -1,9 +1,29 @@
...
+
+def windows_pytest_mark_asyncio(pytest_mark_asyncio):
+    def wrapper(*args, **kwargs):
+        if args and inspect.iscoroutinefunction(args[0]):
+
+            async def method_proxy(*x):
+                await windows_get_result(args[0](*x))
+
+            return pytest_mark_asyncio(method_proxy)
+        return windows_pytest_mark_asyncio(pytest_mark_asyncio(*args, **kwargs))
```
Comment (elacuesta) [inline tests/__init__.py:19]: This looks like an infinite loop, probably only avoided because all test functions are coroutine functions.

File: tests/__init__.py [hunk 3]
```diff
@@ -1,9 +1,29 @@
...
-
+import pytest
 from scrapy import Request
 from scrapy.http.response.html import HtmlResponse
 from scrapy.utils.test import get_crawler
 
+from scrapy_playwright.handler import windows_get_result
+
+
+def windows_pytest_mark_asyncio(pytest_mark_asyncio):
```
Comment (elacuesta) [inline tests/__init__.py:11]: I suspect this will not work after #252, is there another pytest hook that could be leveraged?

**Review by elacuesta — COMMENTED** [review]
Awesome work, thank you very much. I have left some comments but they're mostly about code organization. I will take some more time to review the actual implementation, this change has big implications and I'd like to get a good understanding before moving forward with it.

File: scrapy_playwright/handler.py [hunk 4]
```diff
@@ -39,15 +40,50 @@
     _maybe_await,
 )
 
+# Supporting for Windows
+if sys.platform == "win32" and sys.version_info >= (3, 8):
+    import threading
+
+    class Var:
+        windows_loop = None
+        windows_thread = None
```
[the remark this answers is not in the document]
Comment (sanzenwin) [reply scrapy_playwright/handler.py:49]: [pylint-w0603](https://pylint.readthedocs.io/en/latest/user_guide/messages/warning/global-statement.html)

File: tests/tests_twisted/test_mixed_requests.py [hunk 5]
```diff
@@ -8,6 +10,10 @@
 from tests.mockserver import StaticMockServer
 
 
+@pytest.mark.skipif(
+    platform.system() == "Windows",
+    reason="On Windows, If the task reaches two, it will be blocked.",
```
[the remark this answers is not in the document]
Comment (sanzenwin) [reply tests/tests_twisted/test_mixed_requests.py:15]: On Windows

Works well:
```
tox -e py-twisted -- tests/tests_twisted/test_mixed_requests.py::MixedRequestsTestCase::test_regular_request
tox -e py-twisted -- tests/tests_twisted/test_mixed_requests.py::MixedRequestsTestCase::test_playwright_request
```


Blocked:
```
tox -e py-twisted
```

File: scrapy_playwright/handler.py [hunk 6]
```diff
@@ -39,15 +40,50 @@
...
+    windows_get_asyncio_event_loop = None
+    windows_get_result = None
+    deferred_from_coro = deferred_from_coro_default
 
-logger = logging.getLogger("scrapy-playwright")
+__all__ = [
+    "ScrapyPlaywrightDownloadHandler",
+    "windows_get_asyncio_event_loop",
+    "windows_get_result",
+]
```
[the remark this answers is not in the document]
Comment (sanzenwin) [reply scrapy_playwright/handler.py:82]: The loop should be exposed, such as for tests or other feature.

File: tests/__init__.py (hunk 2 above)
Comment (sanzenwin) [reply tests/__init__.py:19]: Thinking about `@pytest.mark.asyncio(scope="class")`

File: tests/__init__.py (hunk 3 above)
Comment (sanzenwin) [reply tests/__init__.py:11]: If you remove pytest-asyncio, you still need to add some wrapper to every test function, as Windows needs a special process.

File: tests/tests_asyncio/test_page_methods.py [hunk 7]
```diff
@@ -159,15 +159,16 @@ async def test_page_method_screenshot(self):
 
                 png_file.file.seek(0)
                 assert png_file.file.read() == req.meta["playwright_page_methods"]["png"].result
-                assert get_mimetype(png_file) == "image/png"
+                if platform.system() != "Windows":
```
[the remark this answers is not in the document]
Comment (sanzenwin) [reply tests/tests_asyncio/test_page_methods.py:162]: To my understanding, `sys.platform` pays more attention to python's compile features

File: scrapy_playwright/handler.py [hunk 8]
```diff
@@ -39,15 +40,48 @@
...
+        if Var.windows_thread is None:
+            if Var.windows_loop is None:
+                Var.windows_loop = asyncio.WindowsProactorEventLoopPolicy().new_event_loop()
+                asyncio.set_event_loop(Var.windows_loop)
+            if not Var.windows_loop.is_running():
+                Var.windows_thread = threading.Thread(
+                    target=Var.windows_loop.run_forever, daemon=True
+                )
+                Var.windows_thread.start()
+        return Var.windows_loop
```
Comment (Gallaecio) [inline scrapy_playwright/handler.py:61]: According to https://docs.scrapy.org/en/latest/topics/asyncio.html#windows-specific-notes Twisted does not support the Proactor event loop. Is that no longer the case with some new version of Twisted?

File: scrapy_playwright/handler.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (elacuesta) [reply scrapy_playwright/handler.py:80]: They can still be imported, `__all__` is for explicitly defining which names will be imported when doing `import *` https://docs.python.org/3/tutorial/modules.html#importing-from-a-package

File: tests/tests_asyncio/test_page_methods.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (elacuesta) [reply tests/tests_asyncio/test_page_methods.py:162]: My point is, can we use the same source for all checks?

File: scrapy_playwright/handler.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (elacuesta) [reply scrapy_playwright/handler.py:49]: That's fine, these objects are only accessed by name from within the scope of this Windows check. Please also make them underscored. If pylint complains, just add a `# pylint: disable` comment to the line(s).

File: scrapy_playwright/handler.py (hunk 8 above)
Comment (elacuesta) [reply scrapy_playwright/handler.py:61]: According to [the Twisted 23.8.0 docs](https://docs.twisted.org/en/twisted-23.8.0/api/twisted.internet.asyncioreactor.AsyncioSelectorReactor.html) this is still the case:
> ...on Python 3.8 and newer the default event loop is asyncio.WindowsProactorEventLoop which is incompatible with [AsyncioSelectorReactor](https://docs.twisted.org/en/twisted-23.8.0/api/twisted.internet.asyncioreactor.AsyncioSelectorReactor.html).

This seems to be running in a different thread, maybe even bypassing the Twisted reactor. I wonder what are the full implications of this, and how desirable it is. The first question that comes to my mind is: do we still need to set the `TWISTED_REACTOR` Scrapy setting or is that irrelevant with this patch?

File: scrapy_playwright/handler.py [hunk 9]
```diff
@@ -343,10 +377,8 @@ async def _download_request(self, request: Request, spider: Spider) -> Response:
                         "spider": spider,
                         "context_name": context_name,
                         "scrapy_request_url": request.url,
-                        "scrapy_request_method": request.method,
-                        "exception": ex,
-                    },
-                    exc_info=True,
+                        "scrapy_request_method": request.method
+                    }
```
Comment (elacuesta) [inline scrapy_playwright/handler.py:381]: How is this related?

**Comment by ProfoundWanderer** [comment]
Is this pull request still actively being worked out?

**Comment by ProfoundWanderer** [comment]
> I personally am not working on this at the moment. I left a few comments/suggestions/questions in this PR. I will try to find some time to work on this in the near future, but I can't make any promises.
> 
> The best way of using this package on Windows is still under WSL.

I don't mind trying to work on it, I just didn't want to step on any toes. Personally, didn't want to go through WSL hoops if it was possible to get this supported on Windows at all.
